ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -19,210 +19,274 @@
 typedef struct _PNP_NOTIFY_ENTRY
 {
     LIST_ENTRY PnpNotifyList;
-    IO_NOTIFICATION_EVENT_CATEGORY EventCategory;
     PVOID Context;
-    UNICODE_STRING Guid;
-    PFILE_OBJECT FileObject;
     PDRIVER_OBJECT DriverObject;
     PDRIVER_NOTIFICATION_CALLBACK_ROUTINE PnpNotificationProc;
+    union
+    {
+        GUID Guid; // for EventCategoryDeviceInterfaceChange
+        struct // for EventCategoryTargetDeviceChange
+        {
+            PFILE_OBJECT FileObject;
+        };
+    };
+    IO_NOTIFICATION_EVENT_CATEGORY EventCategory;
+    volatile LONG Deleted;
 } PNP_NOTIFY_ENTRY, *PPNP_NOTIFY_ENTRY;
 
-KGUARDED_MUTEX PnpNotifyListLock;
-LIST_ENTRY PnpNotifyListHead;
+KGUARDED_MUTEX IopTargetDeviceNotifyLock;
+KGUARDED_MUTEX IopDeviceClassNotifyLock;
+KGUARDED_MUTEX IopHwProfileNotifyLock;
+
+LIST_ENTRY PiNotifyHwProfileListHead;
+LIST_ENTRY PiNotifyDeviceInterfaceListHead;
```

Comment:
You can use `_Guarded_by_` annotations here to make the relationship to the locks obvious.